### ISSUE TITLE
Don't publish edits immediately (but in batches/on intervals) to protect privacy

### ISSUE THREAD

**Issue by keunes** [organicmaps/organicmaps#1325]
When travelling abroad or walking around and about in my own neighborhood I try to make map changes to increase map accuracy or detail (e.g. opening hours, amenities).

These changes are uploaded to & applied on OSM immediately. Therefore my edits leave a digital trace on OSM (esp my user page) of where I am and have been, including time.

To kinda obfuscate traces it would be great if changes would be uploaded in batches (e.g. of 5) or in time intervals (e.g. every week).

It would also be great if there were some mechanism to pause uploading of edits. This can be useful when travelling, for example, so edits are published only when returned at home. The police advices to do the same for holiday pictures to avoid criminals being able to figure out when the house is 'unguarded'.

Of course the maximum waiting period shouldn't be too long, and also any outstanding changed should be uploaded before new maps are downloaded (or well before, so that the changes are included in the map updates).

**Comment by biodranik** [comment]
The algorithm should be smart, but take into account that:
1. We cannot delay upload for too long to avoid merge conflicts, and to make sure that changes will not be lost (what if you are going into a long offline route?).
2. OSM people do not like when there are many edits in a changeset and there is a big distance between edits. They usually monitor their area and edits in different cities uploaded in one batch trigger different people in between (inside the edit rectangle) to review/validate changes.

It would be great to avoid manual upload control to have it simple and safe for most people. Any ideas are welcome.

**Comment by keunes** [comment]
Thanks a lot @biodranik for those comments.

> We cannot delay upload for too long to avoid merge conflicts

That can happen anyway, right? E.g., between OM users, and between an OM and an online (OSM) editor (the former doesn't know about edits by the latter).

But to avoid conflicts, I think we need two maximums.

I don't submit too many edits, so a period (time limit) of 2 weeks would protect my privacy quite well. There's low risk of conflicts because the low number of data points I'm editing.

However, conflict chance increases as more edits are made. For users that do many edits, I'd suggest we also upload changes if there's 10 edits are waiting.

When upload happens, a 'last upload' date should be updated, to ensure the time-limit is respected (not too short, as that could accidentally lead to the privacy issue that's currently there).

> We cannot delay upload for too long … to make sure that changes will not be lost (what if you are going into a long offline route?).

That's difficult to predict. I'm hardly ever offline (without connection) for a long time, but others (e.g., in remote areas) might be offline most of the time.

I'm also not sure in which cases edits might be 'lost'. There is of course the situation where no more edits can be made to the map because a new map is available. At that point the edits by the user could still be valid – there's just a higher chance of conflicts.

To avoid conflicts I'd propose to upload edits as structured notes if new maps are available. (Similar to structured change proposal notes described in another issue.)

I'd suggest a notification/dialog if upload is due and there's no connection. The user can then temporarily ensure internet access, or look for internet access (if a physical location needs to be reached). Edits waiting to be uploaded, can then be uploaded in the background when an internet connection is available.
If 4 days after the threshold (be it time or number of changes) has been reached they still weren't uploaded, they could be converted to notes, as above.

> OSM people do not like when there are many edits in a changeset and there is a big distance between edits

I wasn't aware of this, but makes sense. Each edit currently is its own changeset, right? Is it feasible/possible to group edits & notes by commune/municipality (with one changeset per municipality)? I guess that'd require the storage of extra metadata, though, for each edit/note.

**Comment by biodranik** [comment]
Thanks! There is another important point to consider: many of our users, especially OSMers, want their changes to be available ASAP because they want to see it online at home, and maybe even correct and update tags which are not editable from OM.

Edits could be lost by reinstalling the app, by losing the phone, breaking it, dropping it from the mountain/into the lake etc. 

In my opinion, 2 weeks delay is too much. Active notifications are also disturbing and it's better to avoid them.

What about an option "Delay OSM uploads for better privacy" option into settings or OSM profile page, and add a random 1-7 days delay?

**Comment by keunes** [comment]
> What about an option "Delay OSM uploads for better privacy" option into settings or OSM profile page, and add a random 1-7 days delay?

Right, such option could work, yeah. If implemented I think it should be enabled by default (for new installs), so the app protects the privacy by default.

I see the appeal of randomness, but this means in some cases your privacy might not be protected. And the thing is: viewers of an online profile won't know that the upload date is random. From larger changesets (which is a logical consequence of longer collection period), however, it is clear that the person cannot have been there at that time (due to the physical constraints around being at multiple locations at one point in time).

> Edits could be lost by reinstalling the app, by losing the phone, breaking it, dropping it from the mountain/into the lake etc.

That last one I very much feared during my current holiday 😁 Though having your phone on aeroplane mode (or being in a remote area w/out connection) also creates that risk.

> There is another important point to consider: many of our users, especially OSMers, want their changes to be available ASAP because they want to see it online at home, and maybe even correct and update tags which are not editable from OM.

Alternatively (to a setting), on the OSM profile page a 'Submit changes and notes now' button could be added. Then no new setting needs to be added ([why fewer options are better](http://neugierig.org/software/blog/2018/07/options.html)). Very active contributors can use the button if they want to see and expand their changes when they get home (if you're out and about and make a change it's not likely you'll _immediately_ check OSM online). Would introduce another button to tap, though.

> In my opinion, 2 weeks delay is too much.

For those power editors that create many edits in different locations, a short timespan is fine.

For me, a delay of one or 4 days doesn't make much sense, though. I submit only two or three edits per month at home. With such small delay, you can still draw conclusions like this: "This week he went to the hairdresser and to the train station to visit city X. The week after he went for an ice-cream in a neighbouring town, probably by bike because there's edits along the route."

When on holiday, I make more edits. If published the same week, it's still obvious I'm not at home at that very time.

Are there any stats on how many people make how many changes through OM? We should be considering both power editors and occasional contributors like me. But I'm curious about the balance as it might (should) impact the decisions/conclusions we would make here.

Remember also that power editors don't risk losing many many edits, as (in addition to the time threshold) I _also_ proposed a number-threshold. Whichever is first applies. So power editors' waiting time would be much shorter.

> Active notifications are also disturbing and it's better to avoid them.

Totally agree. I just proposed it as a backup mechanism for contributors that might not be connected as often as most folks in European and American (urban) areas. (This is not to say that in all other parts of the world there's less connectivity; these are only the areas I am sure to expect connectivity.) Those with high connectivity I wouldn't expect to see the notification.

I should've written the feature in short as follows: 

* If no connection available _even after 24 hours following scheduled upload_, notify the user of outstanding upload (offer option to turn off aeroplane mode, if active)

**Comment by biodranik** [comment]
I don't think that the privacy option should be a default, otherwise it will break existing users' patterns.

Also, OSM editing literally requires that you should see that place and be there. So it is not ideally private by design.

BTW, you can log out of OSM, make any number of edits, and then login back again to upload. I think it should work now as a workaround for the privacy issues.

**Comment by keunes** [comment]
> I don't think that the privacy option should be a default, otherwise it will break existing users' patterns.

It doesn't break existing user's patterns if introduced as a default only for new installs (?) I didn't mean to suggest to change existing installations.

> OSM editing literally requires that you should see that place and be there

Didn't know, but sure, makes sense. Doesn't mean it has to be 'live' though :)

> it should work now as a workaround for the privacy issues

It is a bit cumbersome, but good idea. Will do that for now, thanks.

**Comment by privatemaker** [comment]
> I don't think that the privacy option should be a default, otherwise it will break existing users' patterns.

@biodranik as this is being discussed elsewhere- not all users are the same! An app which markets itself as a "privacy-focused" really SHOULD have defaults which are the most privacy respecting in all circumstances. This is one of those circumstances.

> OSM profile page a 'Submit changes and notes now'

This `batch submit` idea that @keunes proposes is a feature which accommodates users with high privacy concerns who are *also* wanting to *sometimes* contribute to OSM data.

**Comment by biodranik** [comment]
For sure we need to support as many user groups as possible. With our _very_ limited resources we try to focus on the widest audiences first. Contributing to OSM implies that the history of your edits is publicly visible and accessible to everyone. Accepting OSM terms when registering implies that users know that. We can improve transparency there, of course, and add some fuzziness too for upload time. And don't forget about many users who want to see their changes _immediately_, and don't care much about privacy concerns.

**Comment by pastk** [comment]
IMHO an opt-in setting to push changes manually via a button (and optionally review them first if needed similar to the `?edits` command) would be a good solution. Maybe with another option of automatically delaying changes by a couple of weeks.

Defaulting to long or random delays (like a week or so) will raise a lot of questions like "Why I don't see my changes on osm.org? Is it a bug or what?". The default behavior should be logical, simple, obvious.
